create_layout.py

```diff
@@ -91,45 +91,120 @@
 import in_toto.models.link
 import in_toto.models.layout
 
-def create_material_rules(links, index):
+def changes_between_snapshots(before_dict, after_dict):
+  '''Given two 'snapshots' of a file structure -- 'before' and 'after' -- 
+    return a tuple specifying which files have been added, 
+    which have been removed, which have been modified, 
+    and which have remained unchanged. 
+    Both these dictionaries have file
+    names as the keys and their hashes as the values.'''
+
+  unchanged_files = []
+  modified_files = []
+  added_files = []
+  removed_files = []
+  for key in before_dict:
+    if key in after_dict:
+      if before_dict[key] == after_dict[key]:
+        # Matching the hashes to check if file was unchanged
+        unchanged_files.append(key)
+      else:
+        modified_files.append(key)
+    else:
+      removed_files.append(key)
+  for key in after_dict:
+    if key not in before_dict:
+      # Looking for new files
+      added_files.append(key)
+
+  # Returning the snapshot of the new file system
+  return (sorted(unchanged_files), sorted(modified_files), sorted(added_files),
+      sorted(removed_files))
+
+def create_material_rules(changes_of_curr_step, links, index):
```

Comment:
I wonder if it'd be saner to define the `create_material_rules` API to be unaware of the index. It should accept two successive links and return the material rules, no? It can even call `changes_between_snapshots` internally using the links passed. If a link is the first in the supply chain, the "preceding" link can be `None`.

```
def create_material_rules(previous_link, current_link):
  ....
```

Does that make sense? I feel like it would simplify this API since this would be called by some iterative block of code anyway, in this instance `create_layout_from_ordered_links`.

cc @lukpueh